### PR TITLE
make device property always return a device with index

### LINKED ISSUE #4849: Failed to get correct device id inside on_fit_start() or setup()
## 🐛 Bug
In DDP training mode, calling self.device in on_fit_start() or setup() function will return "cuda", while calling self.device in training_step will return "cuda:id" where id corrspondes to the correct device id. 

### To Reproduce
~~~
BoringModel(pl.LightningModule):
    def on_fit_start(self):
        print(self.device)      # will always return "cuda" regardless which device it on
    def setup(self, stage):
        print(self.device)      # will always return "cuda" regardless which device it on
    def training_step(self, **kwargs):
        print(self.device)      # will return device ids, e.g. "cuda:0"
~~~

### Expected behavior
on_fit_start() and/or setup() should be able to get correct device ids as same in training_step().

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #4849 by always returning a device index for cuda devices.

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LTGM !

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 1]
```diff
@@ -37,7 +37,13 @@ def dtype(self, new_dtype: Union[str, torch.dtype]):
 
     @property
     def device(self) -> Union[str, torch.device]:
-        return self._device
+        device = self._device
+
+        # make this more explicit to always include the index
+        if device.type == 'cuda' and device.index is None:
+            return torch.device(f'cuda:{torch.cuda.current_device()}')
```
Comment (awaelchli) [inline pytorch_lightning/utilities/device_dtype_mixin.py:44]: passing the index via string vs. additional argument does not make a difference, right?

**Review by awaelchli — APPROVED** [review]
maybe there is a reason why PyTorch does not set the index explicitly in case of index=None?

**Review by SkafteNicki — APPROVED** [review]
LGTM :)

File: pytorch_lightning/utilities/device_dtype_mixin.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/utilities/device_dtype_mixin.py:44]: No, that's the same.

> maybe there is a reason why PyTorch does not set the index explicitly in case of index=None?

I was thinking the same, but when you do the following:
```python
import torch
a = torch.rand(1, 2, 3)
a = a.cuda()
print(a.device)
```

you get a cuda device with an index, so this part of the parsing logic must be part of the tensor class and not the module class there.

**Review by ydcjeff — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]
